programs/pkey/pk_sign.c

```diff
@@ -100,8 +100,7 @@ int main( int argc, char *argv[] )
 
     if( ( ret = mbedtls_pk_parse_keyfile( &pk, argv[1], "" ) ) != 0 )
     {
-        ret = 1;
-        mbedtls_printf( " failed\n  ! Could not open '%s'\n", argv[1] );
+        mbedtls_printf( " failed\n  ! Could not parse '%s'\n", argv[1] );
```

Comment:
@kevlut: Many thanks for your contribution. I concur with you that hiding the real error code behind the assignment to 1 is inconvenient for debugging. However, the assignment is not unnecessary as `mbedtls_pk_parse_keyfile()` can return integers that are bigger than 8-bit values as required by POSIX. Removing the assignment to 1 and executing the goto exit means that there is a chance the application now returns a non-posix compliant error code to the system. To avoid this problem and still get to see the error code from `mbedtls_pk_parse_keyfile()` I would suggest reversing the order of `ret = 1;` and `mbedtls_printf( " failed\n  ! Could not parse '%s'\n", argv[1] );`.

@kevlut, @AndrzejKurek : what are your thoughts?

Reply:
@andresag01 I agree. It is not a critical issue, but since this is being fixed it is worth a little extra work to have it really right. I also agree with what @hanno-arm has stated here: https://github.com/ARMmbed/mbedtls/pull/532#issuecomment-336706058.

My suggestion to handle it is to teat `ret` as the "error status" throughout the entire program and only convert it to POSIX compliant status at the end of `main()`. The blunt way to do it is to change the return command to:

    return( !!ret );

A more elegant way:

    return( ret ? EXIT_FAILURE : EXIT_SUCCESS );

or an equivalent implemented with an `if` statement.

Reply:
I agree with @k-stachowiak - it's better to keep the appropriate return values for printing and only convert the error code at the end. If not, we'd have to analyze all of the other return codes, for example from mbedtls_ctr_drbg_seed, mbedtls_pk_sign etc., and these may change in the future.
It seems for me that checking for POSIX-compliancy at the end would be the best solution.